databuilder/databuilder/models/type_metadata.py

```diff
@@ -0,0 +1,306 @@
+            yield from self.data_type.create_node_iterator()
+
+    def create_relation_iterator(self) -> Iterator[GraphRelationship]:
+        if not self.start_label:
+            raise Exception('Required start node label cannot be None')
+        if not self.start_key:
+            raise Exception('Required start node key cannot be None')
+
+        if not self.is_terminal_type():
+            yield self._get_relation(self.start_label, self.start_key)
+            yield from self.data_type.create_relation_iterator()
+
+
+class MapTypeMetadata(TypeMetadata):
+    def __init__(self,
+                 key: str,
+                 value: Union[TypeMetadata, str],
+                 start_label: Optional[str] = None,
+                 start_key: Optional[str] = None) -> None:
+        super(MapTypeMetadata, self).__init__(start_label, start_key)
+        self.key = key
+        self.value = value
+
+    def __str__(self) -> str:
+        return f"map<{self.key},{self.value.__str__()}>"
+
+    def is_terminal_type(self) -> bool:
+        return not isinstance(self.value, TypeMetadata)
+
+    def _get_node_key(self, start_key: str) -> str:
+        return f"{start_key}/map"
+
+    def _get_node(self, start_key: str) -> GraphNode:
+        node = GraphNode(
+            key=self._get_node_key(start_key),
+            label=TypeMetadata.TYPE_NODE_LABEL,
+            attributes={
+                TypeMetadata.TYPE_KIND: 'map',
+                TypeMetadata.TYPE_MAP_KEY: self.key.__str__(),
+                TypeMetadata.TYPE_MAP_VALUE: self.value.__str__()
+            }
+        )
+        return node
+
+    def _get_relation(self,
+                      start_label: str,
+                      start_key: str) -> GraphRelationship:
+        relation = GraphRelationship(
+            start_label=start_label,
+            start_key=start_key,
+            end_label=TypeMetadata.TYPE_NODE_LABEL,
+            end_key=self._get_node_key(start_key),
+            type=TypeMetadata.TYPE_RELATION_TYPE,
+            reverse_type=TypeMetadata.INVERSE_TYPE_RELATION_TYPE,
+            attributes={}
+        )
+        return relation
+
+    def create_node_iterator(self) -> Iterator[GraphNode]:
+        if not self.start_key:
+            raise Exception('Required start node key cannot be None')
+
+        if not self.is_terminal_type():
+            yield self._get_node(self.start_key)
+            yield from self.value.create_node_iterator()
+
+    def create_relation_iterator(self) -> Iterator[GraphRelationship]:
+        if not self.start_label:
+            raise Exception('Required start node label cannot be None')
+        if not self.start_key:
+            raise Exception('Required start node key cannot be None')
+
+        if not self.is_terminal_type():
+            yield self._get_relation(self.start_label, self.start_key)
+            yield from self.value.create_relation_iterator()
+
+
+class StructItem():
```

Comment:
having the name ON TypeMetadata would eliminate the need for this. let's discuss.

Reply:
fyi I got rid of StructItem and added the description to the base class as we discussed, but I didn't add the name or type there due to the different way they're represented in a couple places. I had started to add them but it felt like it didn't fully translate to all the subclasses, mostly in the struct class since it contains a dict of names to types rather than just one type. PTAL and we can discuss if you see a better way of making it more generic